File: apps/app/lib/activity.ts
"use client";

import { type Address, type PublicClient, parseAbiItem } from "viem";

import { API_BASE } from "@/lib/api";
import { MARKETS } from "@/lib/venues";

/**
 * What the account itself did, read from its own logs.
 *
 * **Why this exists at all.** "Recent activity" read Aqua movements for the wallet, and Aqua
 * movements are `Pulled`/`Pushed` under a shipped mandate. So the panel sat empty on an account
 * that had just been armed, funded, and had half a dollar supplied to Morpho by the enclave — three
 * things happened and the page said nothing had. It was answering a narrower question than its own
 * title asks.
 *
 * `HelicoAccount` emits an event for each of those three, so the answer is one `eth_getLogs` away
 * and needs no subgraph change. Verified against the live account on 11 September: eight logs,
 * `AgentChanged` twice, `VenuePermitted` four times, and one `IdleCapitalMoved` carrying 490081 of
 * USDC into Morpho.
 *
 * The Aqua movements stay where they are. These are different questions and the page shows both.
 *
 * **The backend answers this now, and the chain is the fallback.** Every browser was scanning from
 * the factory\'s deployment block on every page load — three `eth_getLogs` calls each spanning a
 * million blocks, against a public endpoint everybody shares. `GET /api/activity` owns the history
 * instead: the log is append-only, so a watermark in its database turns the second read into a
 * narrow query. Measured against the live account: 671 ms on the first ask, **0.67 ms** on the next.
 *
 * The direct read stays as the fallback, which is the property `/api/graph` deliberately keeps —
 * this page worked before the backend existed and should not stop working when it is down.
 */
const MOVED = parseAbiItem(
  "event IdleCapitalMoved(address indexed pool, address indexed asset, uint256 amount, bool supplied)",
);
const AGENT = parseAbiItem("event AgentChanged(address indexed agent)");
const VENUE = parseAbiItem(
  "event VenuePermitted(address indexed pool, bool allowed)",
);

export type AccountEvent = {
  /** What the chain said happened, kept beside the sentence so a fold does not read prose. */
  kind: "moved" | "agent" | "venue" | "in" | "out" | "unknown";
  /** Base units of the asset, for the rows that carry one. Zero for the rest. */
  units: bigint;
  /** True for money arriving: supplied to a market, or transferred into the account. */
  into: boolean;
  /** Block first, log index second. Two events in one transaction keep the order they happened. */
  key: string;
  what: string;
  /** The amount, already formatted, for the rows that move money. Empty for the rest. */
  amount: string;
  /** The market, for the rows that name one. Empty for the rest. */
  where: string;
  block: bigint;
  /** Unix seconds, or null when the block's header could not be read. Never invented. */
  at: number | null;
  tx: `0x${string}`;
};

/**
 * A market's name, disambiguated only when it needs to be.
 *
 * Two of the four markets are labelled "Compound v3" — one for USDC and one for ETH — so a list of
 * permissions read "Allowed Compound v3" twice with nothing to tell them apart. True and useless.
 * The asset is appended only for a label that more than one market shares, so the common case stays
 * short.
 */
const marketName = (pool: string) => {
  const market = MARKETS.find(
    (m) => m.pool.toLowerCase() === pool.toLowerCase(),
  );
  if (!market) return `${pool.slice(0, 6)}…${pool.slice(-4)}`;
  const shared = MARKETS.filter((m) => m.label === market.label).length > 1;
  return shared ? `${market.label} (${market.assetSymbol})` : market.label;
};

const amount = (v: bigint | string) =>
  (Number(v) / 1e6).toLocaleString(undefined, {
    minimumFractionDigits: 2,
    maximumFractionDigits: 2,
  });

/**
 * Newest first.
 *
 * `fromBlock` is the factory's own deployment block: nothing can predate the contract that creates
 * these accounts, so it is exact rather than a guess, and it saves scanning the 485 million blocks
 * before it.
 */
export async function readAccountActivity(
  client: PublicClient,
  account: Address,
  fromBlock = 502_979_401n,
): Promise<AccountEvent[]> {
  const served = await fromBackend(account);
  if (served) return served;
  return readFromChain(client, account, fromBlock);
}

/** The shape `GET /api/activity` returns: the fields of the events, not sentences about them. */
type ServedEvent = {
  block: number;
  blockTime?: number;
  logIndex: number;
  kind: "moved" | "agent" | "venue" | "in" | "out" | "unknown";
  tx: string;
  pool?: string;
  amount?: string;
  supplied?: boolean;
  allowed?: boolean;
  agent?: string;
};

/**
 * Null rather than throwing, because a backend that is down is a reason to read the chain, not a
 * reason to show a person an error about a service they did not know existed.
 */
async function fromBackend(account: Address): Promise<AccountEvent[] | null> {
  try {
    const res = await fetch(
      `${API_BASE}/api/activity?account=${account.toLowerCase()}`,
      { headers: { accept: "application/json" } },
    );
    if (!res.ok) return null;
    const body = (await res.json()) as { events?: ServedEvent[] };
    if (!Array.isArray(body.events)) return null;
    return body.events.map((e) => ({
      kind: e.kind,
      units: BigInt(e.amount ?? "0"),
      into: e.kind === "out" ? false : Boolean(e.supplied),
      key: `${e.block}-${e.logIndex}`,
      what: sentence(e),
      amount: carriesMoney(e.kind) ? `${amount(e.amount ?? "0")} USDC` : "",
      where: e.pool ? marketName(e.pool) : "",
      block: BigInt(e.block),
      at: e.blockTime ? e.blockTime : null,
      tx: e.tx as `0x${string}`,
    }));
  } catch {
    return null;
  }
}

const carriesMoney = (kind: ServedEvent["kind"]) =>
  kind === "moved" || kind === "in" || kind === "out";

/**
 * The English, composed here in both paths.
 *
 * The backend returns fields and no prose on purpose: a second set of sentences in Go, free to
 * drift from these, is the thing `graph.go` refuses for the same reason.
 */
function sentence(e: ServedEvent): string {
  switch (e.kind) {
    case "moved":
      // The amount and the market have columns of their own, so the sentence says the verb.
      return e.supplied ? "Put money to work" : "Took money back out";
    case "agent":
      return e.agent === "0x0000000000000000000000000000000000000000"
        ? "Took the agent's permission away"
        : "Named the agent that may move capital";
    case "venue":
      return e.allowed ? "Allowed a market" : "Revoked a market";
    case "in":
      return "Money in";
    case "out":
      return "Money out";
    default:
      return "Something happened on this account";
  }
}

async function readFromChain(
  client: PublicClient,
  account: Address,
  fromBlock: bigint,
): Promise<AccountEvent[]> {
  const [moved, agent, venue] = await Promise.all([
    client.getLogs({ address: account, event: MOVED, fromBlock }),
    client.getLogs({ address: account, event: AGENT, fromBlock }),
    client.getLogs({ address: account, event: VENUE, fromBlock }),
  ]);

  const rows: AccountEvent[] = [
    ...moved.map((l) => ({
      kind: "moved" as const,
      units: l.args.amount ?? 0n,
      into: Boolean(l.args.supplied),
      key: `${l.blockNumber}-${l.logIndex}`,
      what: l.args.supplied ? "Put money to work" : "Took money back out",
      amount: `${amount(l.args.amount ?? 0n)} USDC`,
      where: marketName(String(l.args.pool)),
      block: l.blockNumber ?? 0n,
      // The fallback has no timestamp: `eth_getLogs` does not carry one and fetching a header per
      // block from a browser is the cost this whole path exists to avoid. No date beats a wrong one.
      at: null,
      tx: l.transactionHash as `0x${string}`,
    })),
    ...agent.map((l) => ({
      kind: "agent" as const,
      units: 0n,
      into: false,
      key: `${l.blockNumber}-${l.logIndex}`,
      what:
        String(l.args.agent) === "0x0000000000000000000000000000000000000000"
          ? "Took the agent's permission away"
          : "Named the agent that may move capital",
      amount: "",
      where: "",
      block: l.blockNumber ?? 0n,
      at: null,
      tx: l.transactionHash as `0x${string}`,
    })),
    ...venue.map((l) => ({
      kind: "venue" as const,
      units: 0n,
      into: false,
      key: `${l.blockNumber}-${l.logIndex}`,
      what: l.args.allowed ? "Allowed a market" : "Revoked a market",
      amount: "",
      where: marketName(String(l.args.pool)),
      block: l.blockNumber ?? 0n,
      at: null,
      tx: l.transactionHash as `0x${string}`,
    })),
  ];

  return rows.sort((a, b) =>
    a.block === b.block
      ? b.key.localeCompare(a.key)
      : Number(b.block - a.block),
  );
}

/**
 * The same history with the mechanical halves taken out, for a list a person reads.
 *
 * Supplying to a market is one transaction that emits two things: `IdleCapitalMoved`, which says
 * what happened, and the token transfer that carries it out. Both are needed to work out what the
 * account is worth — one moves the working side, the other the liquid side — and showing both reads
 * as the money leaving twice.
 *
 * The pairing is by transaction and figure rather than by counterparty, because the counterparty is
 * not reliably a market's own address: Aave pulls to its receipt token, not to the pool the event
 * names. A transfer matched by neither stays, which is the case that matters — that is the owner
 * funding the account, or taking it back.
 */
export function withoutVenueLegs(events: AccountEvent[]): AccountEvent[] {
  const legs = new Set(
    events.filter((e) => e.kind === "moved").map((e) => `${e.tx}-${e.units}`),
  );
  return events.filter(
    (e) =>
      (e.kind !== "in" && e.kind !== "out") || !legs.has(`${e.tx}-${e.units}`),
  );
}

/**
 * Which slice of a list one page is, with the page index clamped into range.
 *
 * **Clamped rather than stored-and-reset.** The activity table's row set shrinks — a wallet
 * disconnects, a query refetches shorter, a subgraph answer comes back empty — and an index kept
 * in state then points past the end. The result is a table with no rows, no error and no
 * explanation, which reads as "nothing ever happened" rather than "you are on page 4 of 1". An
 * effect that watched the length and reset it would be a second source of truth about the same
 * number; deriving it cannot get out of step.
 *
 * `pages` is at least 1 so an empty list has a page rather than zero of them, which is what makes
 * `current` safe to use as an index without a second guard at the call site.
 */
export function pageOf(
  total: number,
  page: number,
  size: number,
): { pages: number; current: number; from: number; count: number } {
  const pages = Math.max(1, Math.ceil(total / size));
  const current = Math.min(Math.max(0, page), pages - 1);
  const from = current * size;
  return {
    pages,
    current,
    from,
    count: Math.max(0, Math.min(size, total - from)),
  };
}

File: apps/be/internal/activity/activity.go
// Package activity reads what a Helico account did, from the account's own logs, and remembers it.
//
// **Why this is not a cache.** `/api/graph` in front of the subgraph is a cache: the same question
// gets the same answer for a TTL and then is asked again. An account's log is append-only, so once
// this has read up to block N it never needs to read below N again. What is kept is not a copy of
// an answer, it is the answer so far plus a watermark — which is why the second read of a busy
// account costs one narrow `eth_getLogs` instead of a scan over twenty-three million blocks.
//
// The browser was doing that scan, per visitor, on every page load. Three `eth_getLogs` calls each
// spanning from the factory's deployment to the head, against a public endpoint shared by everyone.
// It worked and it will not keep working.
//
// **What it does not do:** interpret. The rows are the fields of the events, and the sentences are
// composed in the dapp where the copy lives. A second set of English in another language, free to
// drift from the first, is the thing `graph.go` refuses for the same reason.
package activity

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"math/big"
	"net/http"
	"strconv"
	"strings"
	"time"
)

// The three events `HelicoAccount` emits that say something a person would want to read. Keccak of
// the signature, computed once and asserted in the tests against the values the chain actually
// carries — a topic typed from memory is a filter that silently matches nothing.
const (
	TopicIdleCapitalMoved = "0x31b0a7503f02d1d00b2290939640f339abd3c47aa9e41dd6fcc7fa17f2d3cad3"
	TopicAgentChanged     = "0x4a2e63eb36ad3c667a1d8d1b18dfbf37d06f96b46b82b526a855175916515add"
	TopicVenuePermitted   = "0xdae93fbd597062922eea00db27bacf375a602e285a1a9eb57639838a5ac75182"
)

// Kind names an event without repeating its topic downstream.
type Kind string

const (
	KindMoved   Kind = "moved"
	KindAgent   Kind = "agent"
	KindVenue   Kind = "venue"
	KindUnknown Kind = "unknown"
)

// Event is one thing the account did. The fields are the log's, not a sentence about it.
type Event struct {
	Block uint64 `json:"block"`
	// BlockTime is Unix seconds. Fetched once per block and kept for ever: a block's timestamp
	// cannot change, so this is the one field here that never needs re-reading. Zero when the
	// endpoint would not say, which the dapp renders as no date rather than as 1970.
	BlockTime uint64 `json:"blockTime"`
	LogIndex  uint64 `json:"logIndex"`
	Kind      Kind   `json:"kind"`
	Tx        string `json:"tx"`
	// Pool is set for moved and venue. Asset and Amount for moved only.
	Pool   string `json:"pool,omitempty"`
	Asset  string `json:"asset,omitempty"`
	Amount string `json:"amount,omitempty"`
	// Supplied distinguishes a move in from a move out; Allowed a permit from a revoke.
	Supplied bool `json:"supplied,omitempty"`
	Allowed  bool `json:"allowed,omitempty"`
	// Agent is set for agent; the zero address means the permission was taken away.
	Agent string `json:"agent,omitempty"`
}

// ErrBadAddress is an address that is not twenty bytes of hex.
var ErrBadAddress = errors.New("that is not an address")

// Normalise lower-cases an address and refuses anything that is not one. Addresses are a primary
// key here, and `0xAbC…` and `0xabc…` being two rows is a bug that looks like duplicate history.
func Normalise(address string) (string, error) {
	a := strings.ToLower(strings.TrimSpace(address))
	if len(a) != 42 || !strings.HasPrefix(a, "0x") {
		return "", ErrBadAddress
	}
	for _, c := range a[2:] {
		if !strings.ContainsRune("0123456789abcdef", c) {
			return "", ErrBadAddress
		}
	}
	return a, nil
}

// Reader fetches logs. Small enough that a test hands over a struct instead of a server.
type Reader interface {
	Logs(ctx context.Context, address string, from, to uint64) ([]Event, error)
	Head(ctx context.Context) (uint64, error)
	// Times answers the timestamp of each block named, skipping any it cannot.
	Times(ctx context.Context, blocks []uint64) (map[uint64]uint64, error)
}

// RPC reads an EVM endpoint over JSON-RPC.
type RPC struct {
	url  string
	http *http.Client
}

// NewRPC builds a reader against url.
func NewRPC(url string, timeout time.Duration) *RPC {
	return &RPC{url: url, http: &http.Client{Timeout: timeout}}
}

type rpcRequest struct {
	ID      int    `json:"id"`
	JSONRPC string `json:"jsonrpc"`
	Method  string `json:"method"`
	Params  []any  `json:"params"`
}

type rpcError struct {
	Message string `json:"message"`
}

type rawLog struct {
	Address     string   `json:"address"`
	Topics      []string `json:"topics"`
	Data        string   `json:"data"`
	BlockNumber string   `json:"blockNumber"`
	LogIndex    string   `json:"logIndex"`
	TxHash      string   `json:"transactionHash"`
}

func (c *RPC) call(ctx context.Context, method string, params []any, out any) error {
	body, err := json.Marshal(rpcRequest{ID: 1, JSONRPC: "2.0", Method: method, Params: params})
	if err != nil {
		return err
	}
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, c.url, strings.NewReader(string(body)))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/json")
	res, err := c.http.Do(req)
	if err != nil {
		return err
	}
	defer res.Body.Close()
	// Bounded, because an endpoint that answers a wide range can answer with megabytes.
	payload, err := io.ReadAll(io.LimitReader(res.Body, 8<<20))
	if err != nil {
		return err
	}
	var envelope struct {
		Error  *rpcError       `json:"error"`
		Result json.RawMessage `json:"result"`
	}
	if err := json.Unmarshal(payload, &envelope); err != nil {
		return fmt.Errorf("%s: %w", method, err)
	}
	// **The error is returned rather than dropped.** A pruned node answers "missing trie node"
	// with HTTP 200, and a helper that reads that as an empty result reports an account with no
	// history instead of an endpoint that cannot answer.
	if envelope.Error != nil {
		return fmt.Errorf("%s: %s", method, envelope.Error.Message)
	}
	return json.Unmarshal(envelope.Result, out)
}

// Head is the latest block the endpoint knows about.
func (c *RPC) Head(ctx context.Context) (uint64, error) {
	var hex string
	if err := c.call(ctx, "eth_blockNumber", []any{}, &hex); err != nil {
		return 0, err
	}
	return parseHex(hex)
}

// Logs returns the account's three events between from and to, inclusive.
func (c *RPC) Logs(ctx context.Context, address string, from, to uint64) ([]Event, error) {
	var raw []rawLog
	err := c.call(ctx, "eth_getLogs", []any{map[string]any{
		"address":   address,
		"fromBlock": "0x" + strconv.FormatUint(from, 16),
		"toBlock":   "0x" + strconv.FormatUint(to, 16),
		"topics": []any{[]string{
			TopicIdleCapitalMoved, TopicAgentChanged, TopicVenuePermitted,
		}},
	}}, &raw)
	if err != nil {
		return nil, err
	}
	events := make([]Event, 0, len(raw))
	for _, l := range raw {
		e, ok := decode(l)
		if ok {
			events = append(events, e)
		}
	}
	return events, nil
}

// Times reads one block header per number, which is the only way to get a log's timestamp:
// `eth_getLogs` does not carry one. Called once per block ever, because the answer is immutable
// and the row it fills is written to the database beside the event.
func (c *RPC) Times(ctx context.Context, blocks []uint64) (map[uint64]uint64, error) {
	out := make(map[uint64]uint64, len(blocks))
	for _, b := range blocks {
		var header struct {
			Timestamp string `json:"timestamp"`
		}
		// `false` asks for hashes rather than whole transactions: a busy block is megabytes of
		// data for one field.
		if err := c.call(ctx, "eth_getBlockByNumber",
			[]any{"0x" + strconv.FormatUint(b, 16), false}, &header); err != nil {
			// One block that will not answer is a missing date, not a failed read.
			continue
		}
		if t, err := parseHex(header.Timestamp); err == nil {
			out[b] = t
		}
	}
	return out, nil
}

func parseHex(s string) (uint64, error) {
	return strconv.ParseUint(strings.TrimPrefix(s, "0x"), 16, 64)
}

// word reads the nth 32-byte word of a hex data field.
func word(data string, n int) string {
	d := strings.TrimPrefix(data, "0x")
	if len(d) < (n+1)*64 {
		return ""
	}
	return d[n*64 : (n+1)*64]
}

func addressFromWord(w string) string {
	if len(w) != 64 {
		return ""
	}
	return "0x" + strings.ToLower(w[24:])
}

// decode turns one log into an Event, or says it is not one of ours.
//
// The indexed parameters are topics and the rest is data, so each field is read from where the ABI
// puts it rather than from a position in a flat list. A log whose shape does not match is dropped:
// this filters on topic0, and a contract that reuses one of these signatures with different
// parameters would otherwise be read as if it were an account of ours.
func decode(l rawLog) (Event, bool) {
	if len(l.Topics) == 0 {
		return Event{}, false
	}
	block, err := parseHex(l.BlockNumber)
	if err != nil {
		return Event{}, false
	}
	index, err := parseHex(l.LogIndex)
	if err != nil {
		return Event{}, false
	}
	e := Event{Block: block, LogIndex: index, Tx: strings.ToLower(l.TxHash)}
	switch strings.ToLower(l.Topics[0]) {
	case TopicIdleCapitalMoved:
		if len(l.Topics) != 3 {
			return Event{}, false
		}
		amount, ok := new(big.Int).SetString(strings.TrimPrefix(word(l.Data, 0), ""), 16)
		if !ok {
			return Event{}, false
		}
		e.Kind = KindMoved
		e.Pool = addressFromWord(strings.TrimPrefix(l.Topics[1], "0x"))
		e.Asset = addressFromWord(strings.TrimPrefix(l.Topics[2], "0x"))
		e.Amount = amount.String()
		e.Supplied = strings.HasSuffix(word(l.Data, 1), "1")
	case TopicAgentChanged:
		if len(l.Topics) != 2 {
			return Event{}, false
		}
		e.Kind = KindAgent
		e.Agent = addressFromWord(strings.TrimPrefix(l.Topics[1], "0x"))
	case TopicVenuePermitted:
		if len(l.Topics) != 2 {
			return Event{}, false
		}
		e.Kind = KindVenue
		e.Pool = addressFromWord(strings.TrimPrefix(l.Topics[1], "0x"))
		e.Allowed = strings.HasSuffix(word(l.Data, 0), "1")
	default:
		return Event{}, false
	}
	return e, true
}
